tdms/src/iterator.cpp

```diff
@@ -411,117 +409,26 @@ void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[]) {
 
   if (nlhs != 31) { throw runtime_error("31 outputs required. Had " + to_string(nlhs)); }
 
-  /*Get fdtdgrid*/
-  if (!mxIsStruct(prhs[input_counter])) {
-    throw runtime_error("Argument " + to_string(input_counter) + " was expected to be a structure");
-  }
+  /*Get fdtdgrid */
+  assert_pointer_is_struct(prhs[input_counter], "fdtdgrid argument "+to_string(input_counter));
   init_grid_tensors(prhs[input_counter], E_s, H_s, materials);
   int I_tot = E_s.I_tot, J_tot = E_s.J_tot, K_tot = E_s.K_tot;
   input_counter++;
-
-  /*Got fdtdgrid*/
-
-
-
-
-
-
-
-
   //fprintf(stderr,"Got fdtdgrid\n");
-  /*Get Cmaterials */
-  if (mxIsStruct(prhs[input_counter])) {
-    num_fields = mxGetNumberOfFields(prhs[input_counter]);
-    //check that all fields are present
-    if (num_fields != 9) {
-      throw runtime_error("Cmaterials should have 9 members, it has " + to_string(num_fields));
-    }
-
-    for (int i = 0; i < 9; i++) {
-      element = mxGetField((mxArray *) prhs[input_counter], 0, Cmaterial_elements[i]);
-      string element_name = Cmaterial_elements[i];
-      ndims = mxGetNumberOfDimensions(element);
-      if (ndims == 2) {
-        dimptr_out = mxGetDimensions(element);
-        if (dimptr_out[0] != 1) {
-          throw runtime_error("Incorrect dimension on Cmaterial: " + element_name);
-        }
-        if (are_equal(Cmaterial_elements[i], "Cax")) {
-          Cmaterial_Cax = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Cay")) {
-          Cmaterial_Cay = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Caz")) {
-          Cmaterial_Caz = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Cbx")) {
-          Cmaterial_Cbx = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Cby")) {
-          Cmaterial_Cby = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Cbz")) {
-          Cmaterial_Cbz = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Ccx")) {
-          Cmaterial_Ccx = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Ccy")) {
-          Cmaterial_Ccy = mxGetPr(element);
-        } else if (are_equal(Cmaterial_elements[i], "Ccz")) {
-          Cmaterial_Ccz = mxGetPr(element);
-        } else {
-          throw runtime_error("element Cmaterial.%s not handled " + element_name);
-        }
-      } else {
-        throw runtime_error("Incorrect dimension on Cmaterial.Ca");
-      }
-    }
-    input_counter++;
-  } else {
-    throw runtime_error("Argument %d was expected to be a structure " + to_string(input_counter));
-  }
-  /*Got Cmaterials */
 
+  /*Get Cmaterials */
+  assert_pointer_is_struct(prhs[input_counter], "Cmaterials argument "+to_string(input_counter));
+  init_cmaterial(prhs[input_counter], Cmaterial);

```

Comment:
Note for reviewers: this is a very bad constructor.. has been fixed in a later commit